### PR TITLE
Improve conversion of property info for the UI

### PR DESCRIPTION
**Pull request by hunger**
A big PR, but a lot of this is testing :-)

* More type-safe API (although it is still not *nice*)
* Also accept locally defined enums. This used to work only with "our" enums...
* Tests!!!

This PR also does *babysteps* towards changing the UI of the property editor, but it is 98% about the backend, and only the parts we had covered before. I still need to add in length, angle, ...  That's next.

The idea is to have something in place for Florian to play with tomorrow.

### CODE REVIEW

File: tools/lsp/preview.rs [hunk 1]
```diff
@@ -358,60 +358,92 @@ fn evaluate_binding(
     drop_location::workspace_edit_compiles(&document_cache, &edit).then_some(edit)
 }
 
+fn convert_simple_string(input: slint::SharedString) -> String {
+    format!("\"{}\"", input.to_string())
```
Comment (ogoffart) [inline tools/lsp/preview.rs:362]: Don't the string needs to be escaped?

File: internal/compiler/parser.rs [hunk 2]
```diff
@@ -916,6 +956,13 @@ impl NodeOrToken {
             NodeOrToken::Token(t) => t.text_range(),
         }
     }
+
+    pub fn next_sibling_or_token(&self) -> Option<Self> {
```
Comment (ogoffart) [inline internal/compiler/parser.rs:960]: where is this used?

File: tools/lsp/preview/ui.rs [hunk 3]
```diff
@@ -234,147 +236,129 @@ fn map_property_declaration(
...
+    let mut value_string = String::new();
+    let mut visual_items = Rc::new(VecModel::default()).into();
+    let mut is_translatable = false;
+    let mut tr_context = String::new();
+    let mut tr_plural = String::new();
+    let mut default_selection = 0;
+    let code =
+        code_block_or_expression.as_ref().map(|cbe| cbe.text().to_string()).unwrap_or_default();
+
+    match property_type {
```
Comment (ogoffart) [inline tools/lsp/preview/ui.rs:277]: Can you not match on the langtype::Type here?

File: tools/lsp/preview/ui.rs [hunk 4]
```diff
@@ -234,147 +236,129 @@ fn map_property_declaration(
...
-                result.push(c);
+            } else if code.is_empty() {
+                kind = PropertyValueKind::String;
             }
-            _ => {
-                result.push(c);
-                if in_unicode_escape {
-                    return None;
+        }
+        e if e.starts_with("enum ") => {
```
Comment (ogoffart) [inline tools/lsp/preview/ui.rs:311]: you'd get the proper enum from Type instead of mapping to a string.

File: tools/lsp/preview.rs (hunk 1 above)
Comment (hunger) [reply tools/lsp/preview.rs:362]: Probably yes:-)

Do we have a helper for that?

File: tools/lsp/preview/ui.rs (hunk 3 above)
Comment (hunger) [reply tools/lsp/preview/ui.rs:277]: That is probably a good idea.

File: internal/compiler/parser.rs (hunk 2 above)
Comment (hunger) [reply internal/compiler/parser.rs:960]: That indeed seems unused now. I'll double check when I am done with the whole thing...

I only added a method here when I needed, but I also changed the user side a lot over time...

Comment (hunger) [reply internal/compiler/parser.rs]: Removed again... incl. the code it calls on `SyntaxToken` and `SyntaxNode`.

File: tools/lsp/preview.rs (hunk 1 above)
Comment (hunger) [reply tools/lsp/preview.rs]: I used `std::escape_debug` as recommended by @ogoffart .

**Comment by hunger** [comment]
I was using `next_sibling_or_token` in my debug printer code that I stripped out before pushing this :-)

File: tools/lsp/preview/ui.rs [hunk 5]
```diff
@@ -234,147 +239,114 @@ fn map_property_declaration(
...
     }
 
-    SimpleValueData {
-        widget: SharedString::new(),
-        meta_data: Rc::new(VecModel::default()).into(),
-        visual_items: Rc::new(VecModel::default()).into(),
+    PropertyValue {
+        kind,
+        value_bool,
+        value_string: value_string.into(),
```
Comment (ogoffart) [inline tools/lsp/preview/ui.rs:341]: (you could use SharedString directly so you don't need to allocate a String.)

File: tools/lsp/preview/ui.rs [hunk 6]
```diff
@@ -234,147 +239,114 @@ fn map_property_declaration(
...
+        .TrContext()
+        .and_then(|n| n.child_text(SyntaxKind::StringLiteral))
+        .and_then(|s| i_slint_compiler::literals::unescape_string(&s))
+        .unwrap_or_default();
+    let plural = tr_node
+        .TrPlural()
+        .and_then(|n| n.child_text(SyntaxKind::StringLiteral))
+        .and_then(|s| i_slint_compiler::literals::unescape_string(&s))
+        .unwrap_or_default();
+    Some((context, plural, text))
```
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:256]: Will be in the next push (with test).

File: tools/lsp/preview/ui.rs [hunk 7]
```diff
@@ -234,147 +239,114 @@ fn map_property_declaration(
...
+            default_selection = i32::try_from(enumeration.default_value).unwrap_or_default();
+            visual_items = Rc::new(VecModel::from(
+                enumeration.values.iter().map(SharedString::from).collect::<Vec<_>>(),
+            ));
+
+            if let Some(expression) = expression {
+                if let Some(text) = expression.child_node(SyntaxKind::QualifiedName).and_then(|n| {
+                    n.text()
+                        .to_string()
+                        .strip_prefix(&format!("{value_string}."))
```
[the remark this answers is not in the document]
Comment (hunger) [reply tools/lsp/preview/ui.rs:323]: Oh, cool, that will be in the next push (with a test).

File: tools/lsp/preview/ui.rs (hunk 5 above)
Comment (hunger) [reply tools/lsp/preview/ui.rs:341]: I was considering that, but found it a bit cumbersome as I update that value in several places and SharedString is less nice to use.

I'll leave that out for now.

**Comment by hunger** [comment]
@ogoffart: Those comments made the PR better, thanks!

**Review by ogoffart — APPROVED** [review]

**Pull request merged**